### PR TITLE
arch/arm/src/imxrt: NXP USB Host driver

### PR DESCRIPTION
**Pull request by sijupunnnoose**
-Add support for USB host driver, initial driver to support TizenRT
host driver stack. This commit is currently been tested only with
camera driver(uvc).

### CODE REVIEW

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 1]
```diff
@@ -2666,7 +2662,7 @@ static int imxrt_qh_ioccheck(struct imxrt_qh_s *qh, uint32_t **bp, void *arg)
 		 */
 
 		**bp = qh->hw.hlp;
-		arch_flush_dcache((uintptr_t)*bp, (uintptr_t)*bp + sizeof(uint32_t));
+		arch_flush_dcache((uintptr_t) * bp, (uintptr_t) * bp + sizeof(uint32_t));
```
Comment (jeongchanKim) [inline os/arch/arm/src/imxrt/imxrt_ehci.c:2665]: As I know, we will not consider the coding rule for arch folder. But I think this is not good to read the code. Please remove the space for pointer operation.

File: os/arch/arm/src/imxrt/usb/host/usb_host_hci.c [hunk 2]
```diff
@@ -0,0 +1,944 @@
+/*
```
Comment (jeongchanKim) [inline os/arch/arm/src/imxrt/usb/host/usb_host_hci.c:1]: Please add Samsung License.

File: os/arch/arm/src/imxrt/imxrt_ehci.h [hunk 3]
```diff
@@ -113,7 +113,7 @@ extern "C" {
 
 #if defined(CONFIG_IMXRT_USBOTG) && defined(CONFIG_USBHOST)
 struct usbhost_connection_s;
-FAR struct usbhost_connection_s *imxrt_ehci_initialize(int controller);
+int imxrt_ehci_initialize(int controller);
```
[the remark this answers is not in the document]
Comment (sijupunnnoose) [reply os/arch/arm/src/imxrt/imxrt_ehci.h:116]: Done

File: os/arch/arm/src/imxrt/imxrt_ehci.c [hunk 4]
```diff
@@ -698,9 +695,9 @@ static const struct imxrt_ehci_trace_s g_trace2[TRACE2_NSTRINGS] = {
 static uint16_t imxrt_read16(const uint8_t *addr)
 {
 #ifdef CONFIG_ENDIAN_BIG
-	return (uint16_t)addr[0] << 8 | (uint16_t)addr[1];
+	return (uint16_t) addr[0] << 8 | (uint16_t) addr[1];
```
[the remark this answers is not in the document]
Comment (sijupunnnoose) [reply os/arch/arm/src/imxrt/imxrt_ehci.c]: Done

File: os/arch/arm/src/imxrt/imxrt_ehci.c (hunk 1 above)
Comment (sijupunnnoose) [reply os/arch/arm/src/imxrt/imxrt_ehci.c]: Done

File: os/arch/arm/src/imxrt/usb/host/usb_host_hci.c (hunk 2 above)
Comment (sijupunnnoose) [reply os/arch/arm/src/imxrt/usb/host/usb_host_hci.c:1]: @jeongchanKim, currently none of IMXRT arch/arm files has Samsung License header. I dont know why, so I would like to follow the same. Please suggest if you further comments on it.

File: os/arch/arm/src/imxrt/imxrt_usbhost.c [hunk 5]
```diff
@@ -194,8 +192,25 @@ int imxrt_usbhost_initialize(void)
 
 	/* Then get an instance of the USB EHCI interface. */
 
-	g_ehciconn = imxrt_ehci_initialize(0);
-	if (!g_ehciconn) {
+#ifdef CONFIG_USBHOST_HUB
+	/* Initialize USB hub class support */
+	if (usbhost_hub_initialize() != OK) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_usbhost.c:197]: you already defined ret, so line 197 & 202 & 213 can be aligned with line 207~210(use ret & add ret to debug message)

Comment (sijupunnnoose) [reply os/arch/arm/src/imxrt/imxrt_usbhost.c]: Corrected

Comment (sijupunnnoose) [reply os/arch/arm/src/imxrt/imxrt_usbhost.c]: Corrected!

**Review by Taejun-Kwon — APPROVED** [review]

**Pull request merged**